xds/internal/xdsclient/transport/ads/ads_stream.go

```diff
@@ -0,0 +1,801 @@
+
+	ctx, cancel := context.WithCancel(context.Background())
+	s.cancel = cancel
+	go s.runner(ctx)
+	return s
+}
+
+// Stop blocks until the stream is closed and all spawned goroutines exit.
+func (s *Stream) Stop() {
+	s.cancel()
+	s.requestCh.Close()
+	<-s.runnerDoneCh
+	s.logger.Infof("Stopping ADS stream")
+}
+
+// Subscribe subscribes to the given resource. It is assumed that multiple
+// subscriptions for the same resource is deduped at the caller. A discovery
+// request is sent out on the underlying stream for the resource type when there
+// is sufficient flow control quota.
+func (s *Stream) Subscribe(typ xdsresource.Type, name string) {
+	if s.logger.V(2) {
+		s.logger.Infof("Subscribing to resource %q of type %q", name, typ.TypeName())
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	state, ok := s.resourceTypeState[typ]
+	if !ok {
+		// An entry in the type state map is created as part of the first
+		// subscription request for this type.
+		state = &resourceTypeState{
+			subscribedResources: make(map[string]*ResourceWatchState),
+			bufferedRequests:    make(chan struct{}, 1),
+		}
+		s.resourceTypeState[typ] = state
+	}
+
+	// Create state for the newly subscribed resource. The watch timer will
+	// be started when a request for this resource is actually sent out.
+	state.subscribedResources[name] = &ResourceWatchState{State: ResourceWatchStateStarted}
+
+	// Send a request for the resource type with updated subscriptions.
+	s.requestCh.Put(typ)
+}
+
+// Unsubscribe cancels the subscription to the given resource. It is a no-op if
+// the given resource does not exist. The watch expiry timer associated with the
+// resource is stopped if one is active. A discovery request is sent out on the
+// stream for the resource type when there is sufficient flow control quota.
+func (s *Stream) Unsubscribe(typ xdsresource.Type, name string) {
+	if s.logger.V(2) {
+		s.logger.Infof("Unsubscribing to resource %q of type %q", name, typ.TypeName())
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	state, ok := s.resourceTypeState[typ]
+	if !ok {
+		return
+	}
+
+	rs, ok := state.subscribedResources[name]
+	if !ok {
+		return
+	}
+	if rs.ExpiryTimer != nil {
+		rs.ExpiryTimer.Stop()
+	}
+	delete(state.subscribedResources, name)
+
+	// Send a request for the resource type with updated subscriptions.
+	s.requestCh.Put(typ)
+}
+
+// runner is a long-running goroutine that handles the lifecycle of the ADS
+// stream. It spwans another goroutine to handle writes of discovery request
+// messages on the stream. Whenever an existing stream fails, it performs
+// exponential backoff (if no messages were received on that stream) before
+// creating a new stream.
+func (s *Stream) runner(ctx context.Context) {
+	defer close(s.runnerDoneCh)
+
+	go s.send(ctx)
+
+	runStreamWithBackoff := func() error {
+		stream, err := s.transport.CreateStreamingCall(ctx, "/envoy.service.discovery.v3.AggregatedDiscoveryService/StreamAggregatedResources")
```

Comment:
Optional (mainly subjective understanding, but I do feel a little strongly from my point of view): As per my other comments and replies, I think that this confuses me. This conflates the full "ads stream", which to me is almost a virtual stream, as it is backed by a given actual stream here, with the actual underlying stream, which could fail. Similar how we differentiate between the top level Channel and SubChannel, and we differentiate between the top level RPC object, and an RPC Attempt.

I was thinking *ADS*Stream and stream, but then it would stutter with the package name.

Another option would be Stream/streamAttempt maybe? Or streamIter?

Reply:
This is just a local variable. We don't have a type that maintains state and provides functionality on a per-stream attempt basis like what have for an RPC attempt on the channel. So, I feel renaming this variable to `streamAttempt` does not provide much value.

Reply:
I renamed the top-level `ads.Stream` to `ads.StreamImpl` and added a more descriptive docstring for it. Let me know how you feel about it now.

Reply:
Yup that looks good thanks :).